### PR TITLE
Code Diff Viewer, version switching on Autograder Output

### LINKED ISSUE #1743: Submission version diff viewer
Ability to view differences between 2 submissions side-by-side.
Reference: https://diff2html.xyz/

### LINKED ISSUE #2090: Allow version switching on autograding output
As a result of #1578, version switching is disabled when viewing autograding output.

However, it might still be preferable to allow version switching (between autograder outputs).

But, if a submission doesn't have an autograder output, instead of skipping it, perhaps display a message and/or option to regrade the submission.
(Given that autograder output is rendered as text, maybe display a message and add a button to the speedgrader?)

### PR DESCRIPTION
**Pull request by damianhxy**
## Description

- Add a code diff viewer tab to golden layout
- All submissions now display an "Autograder Output" file for uniformity
- Added version switching for Autograder Output
- "Run Autograder" button added when autograder defined and viewing autograding output
- Fix check for detecting if announcements exist
- Tweak heights to ensure that speedgrader is usable with and without announcements
- Remove dead bootstrap tooltip code
- Add titles to access code buttons
- Refactor submissions controller

## Motivation and Context

Closes #2090 
Closes #1743

### Version switching on autograder output

Now able to switch between versions when viewing autograder output. Will always go to the immediate next / previous version, with a message if the output does not exist.

**Output exists (and autograder exists)**
Screenshot 2024-02-17 at 17 04 28
Note the "Run Autograder" button

**Output does not exist, autograder exists**
Screenshot 2024-02-17 at 17 04 36
Note the message, and the "Run Autograder" button

**Output does not exist, autograder does not exist**
Screenshot 2024-02-17 at 17 04 49
Note the message, and NO "Run Autograder" button

### Code Diff Viewer

**Text file - changes exist**
Screenshot 2024-02-18 at 00 57 37

**Text file - no changes**
Screenshot 2024-02-18 at 00 57 40

**Version 1 of a file**
Screenshot 2024-02-18 at 00 57 08

**Other error, e.g. viewing a PDF or Binary file**
Screenshot 2024-02-18 at 00 57 23

**Autograder Output**
Screenshot 2024-02-18 at 00 58 32

**Large files**
Screenshot 2024-02-18 at 00 56 55

### Height tweaks
- In normal use, there is no longer any extraneous scrollbar at the side. Content of all panels (e.g. code viewer, diff viewer) are fully accessible through scrolling if necessary.
- In the presence of announcements, the speedgrader is made "shorter" accordingly, but all content is still accessible through scrolling if necessary.

**Before (CMU prod)**
Screenshot 2024-02-18 at 01 00 52

**After**
Screenshot 2024-02-18 at 01 00 57

**With announcements**
Screenshot 2024-02-18 at 01 02 24

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] I have run rubocop and erblint for style check. If you haven't, run `overcommit --install && overcommit --sign` to use pre-commit hook for linting
- [ ] My change requires a change to the documentation, which is located at [Autolab Docs](https://docs.autolabproject.com/)
- [ ] I have updated the documentation accordingly, included in this PR

### CODE REVIEW

**Comment by KesterTan** [comment]
I think it would be better if the whole line is new, to just remove the highlighting and have the line highlighted light green.

For eg, in this case, maybe just the light green highlight without the dark green.

Screenshot 2024-02-24 at 2 58 25 AM

**Comment by KesterTan** [comment]
Also, I am a little confused about the selective strong highlighting. If these are 2 extremely different files so shouldn't everything be light green?

Screenshot 2024-02-24 at 3 00 23 AM

File: app/views/submissions/_diff_viewer.html.erb [hunk 1]
```diff
@@ -0,0 +1,31 @@
+<div id="diff-box">
+  <% if @prevVersion %>
+    <% title = "Diff with previous version of #{@displayFilename} (version #{@prevVersion[:version]})" %>
+    <% diff_header = "--- #{title}\n+++ #{title}\n" %>
+    <% begin %>
+    <% diff_str = Diffy::Diff.new(@prev_file_contents, @file_contents, include_diff_info: true).to_s(:text) %>
+    <% if diff_str.present? %>
```
Comment (KesterTan) [inline app/views/submissions/_diff_viewer.html.erb:7]: small nit but maybe we could condense this

<% diff_str = diff_str.present? ? [diff_header] + diff_str.lines.slice(2..-1)&.join("") : diff_header %>

**Comment by KesterTan** [comment]
Personally prefer non-synchronized scrolling on both files, especially if file sizes are pretty different.

**Comment by damianhxy** [comment]
> I think it would be better if the whole line is new, to just remove the highlighting and have the line highlighted light green.

Seems to be a limitation of the library, as I don't see a way to specify a "line similarity score threshold".

In general, there is no semantic info associated with a diff, so the library can and will match stuff like whitespace, brackets, and parenthesis.

> Personally prefer non-synchronized scrolling on both files, especially if file sizes are pretty different.

Discussed offline as a won't change, since it would make comparisons more tedious in general and doesn't really bring any benefit.

**Review by KesterTan — APPROVED** [review]
LGTM

**Pull request merged**